Adjust documentation of extension types: Cast/match does not run cons… (#6813)
dart-lang/site-www: src/content/language/extension-types.md

void main() {
  var n = NumberE(1);

>   // The run-time type of 'n' is the representation type 'int'.

Reviewer: ```suggestion
  // The runtime type of 'n' is the representation type 'int'.
```

Author: I think it does make sense to keep using 'run-time' here.

Here are the three related words/phrases as I usually classify them: 'Run time' can be used as a noun, denoting the points in time where a given program is being executed ("the error is not detected statically, it only emerges at run time"). 'Run-time' can be used as an adjective, associating some entity with run time (".. so it's a run-time error"). 'Runtime' is a software entity which is used to support the execution of a program; it includes elements like primitive operations that can't be expressed in the language itself, memory management and garbage collection, and other support facilities (".. and the behavior whereby the stack is unrolled and an exception is thrown is handled by the runtime.").

I think it's a pretty standard choice to express noun phrases as separate words and adjectives with hyphens, so that one should not be controversial.

The choice to use 'runtime' as yet another word with its own meaning is not so easy to support with explicit dictionary definitions, but pages like https://learn.microsoft.com/en-us/dotnet/core/diagnostics/built-in-metrics-runtime, https://docs.rs/tokio/latest/tokio/runtime/struct.RuntimeMetrics.html, https://docs.oracle.com/en-us/iaas/jms/doc/java-runtime-metrics.html, and https://pkg.go.dev/runtime/metrics show that at least the .NET, Rust, Java, and Go communities do use 'runtime' to denote the entity that supports program execution (and they also mention garbage collection and similar tasks).

>   if (n is int) print(n); // Prints 1.

Author: I changed `n.value` to `n` because it has the same behavior, and `n.value` raises a question which is slightly tricky to explain, and it's not the topic of this text.

The tricky part is that `n` is not promoted to `int` even though we have just tested that `n is int` and the result was `true`. This means that `n` still has the static type `NumberE` when we reach the expression `n.value`. This might very well be confusing to a reader who does not consider the fact that `int` is not a subtype of `NumberE`. A reader who thinks that the "promotion" has occurred will not understand why `n.value` isn't a compile-time error. With `print(n)`, the code works whether or not `n` has been promoted, so we just avoid talking about promotion entirely.

Reviewer: Sounds good to me. @parlough any thoughts on this?

}
```

When `i` gets the static type `NumberE` based on a cast or a pattern match, `i`
still refers to the same object, and `v` refers to the same object as `i`, and

> nothing has changed about this object. It is only the static type that changed

Reviewer: ```suggestion
the object itself remains unchanged. It is only the static type that changed
```

(and hence the methods that we can call on this object). In particular, the
change of type does not involve an invocation of a constructor. If you wish to
execute a constructor (for example, to perform some kind of validation) then it

> is necessary to write an explicit constructor invocation (such as `NumberE(i)`).

Reviewer 2: Thanks for adding this additional clarification. I've seen incorrect assumptions around this behavior a few times, so I'm glad we call out it specifically now.
